uring/src/main/scala/fs2/io/uring/net/UringDatagramSocket.scala

```diff
@@ -0,0 +1,145 @@
+package fs2
+package io
+package uring
+package net
+
+import cats.effect.kernel.Async
+import cats.effect.kernel.Resource
+import cats.effect.kernel.Sync
+import cats.effect.std.Mutex
+import cats.syntax.all._
+import com.comcast.ip4s.IpAddress
+import com.comcast.ip4s.SocketAddress
+import fs2.Pipe
+import fs2.io.net.Datagram
+import fs2.io.net.DatagramSocket
+import fs2.io.uring.unsafe.uring._
+import fs2.io.uring.unsafe.util._
+
+import scala.scalanative.libc.stdlib._
+import scala.scalanative.libc.errno._
+import scala.scalanative.posix.sys.socket._
+import scala.scalanative.posix.sys.socketOps._
+import scala.scalanative.posix.sys.uio
+// import scala.scalanative.posix.sys.uioOps._ Implementation of iovecOps is missing
+import scala.scalanative.unsafe._
+import scala.scalanative.unsigned._
+import com.comcast.ip4s.MulticastJoin
+
+private[net] final class UringDatagramSocket[F[_]](
+    ring: Uring[F],
+    fd: Int,
+    buffer: ResizableBuffer[F],
+    defaultReadSize: Int,
+    readMutex: Mutex[F],
+    writeMutex: Mutex[F],
+    val iov: Ptr[uio.iovec],
+    val msg: Ptr[msghdr],
+    val remoteAddr: Ptr[sockaddr_storage]
+)(implicit F: Async[F])
+    extends DatagramSocket[F] {
+
+  private[this] def recvmsg(msg: Ptr[msghdr], flags: Int): F[Int] =
+    ring.call(io_uring_prep_recvmsg(_, fd, msg, flags))
+
+  def read: F[Datagram] =
+    readMutex.lock.surround {
+      for {
+        buf <- buffer.get(defaultReadSize)
+
+        _ <- F.delay {
+          iov._1 = buf
+          iov._2 = defaultReadSize.toULong
+        }
+
+        _ <- F.delay {
+          msg.msg_name = remoteAddr.asInstanceOf[Ptr[Byte]]
+          msg.msg_namelen = sizeof[sockaddr_storage].toUInt
+          msg.msg_iov = iov
+          msg.msg_iovlen = 1
+          msg.msg_control = null
+          msg.msg_controllen = 0.toUInt
+          msg.msg_flags = 0
+        }
+
+        readed <- recvmsg(msg, 0)
+
+        remoteAddress <- F.delay {
+          SocketAddressHelpers
+            .toSocketAddress(remoteAddr.asInstanceOf[Ptr[sockaddr]])
+            .getOrElse(throw new IOException("Failed to get remote address"))
+        }
```

Comment:
Here you don't need to throw an exception to raise an error. You can directly convert from Either[Throwable, A] to F[A] with `F.fromEither`. I'm not sure we need to suspend, since this is seems just a conversion in memory, here @armanbilge can help more than I can.
```suggestion
        remoteAddress <- F.fromEither(
          SocketAddressHelpers
            .toSocketAddress(remoteAddr.asInstanceOf[Ptr[sockaddr]])
        )
```

Reply:
Yes exactly as Lorenzo suggests 💯  and right, no need to suspend here I think since there is no side-effect.

I like using `F.fromEither(x)` and there is also a syntax `x.liftTo[F]` as well.

Reply:
Oh I see thanks Lorenzo and Arman for the feedback :)